### PR TITLE
Fix inconsistent source code hashing behavior with md5sum

### LINKED ISSUE #139: SourceCode Annotator hashing is inconsistent with md5sum shell command
According to #124 and #125, the checksum hashing algorithm for the source code annotator is:

1. Generate a list of files and sort them
2. Generate the hash of the files and add them as a space-separated suffix to the file path
e.g.
```
acbd18db4cc2f85cedef654fccc4a4d8  /tmp/junit11145264397678655906/sourceCode/file1
ae3e83e2fab3a7d8683d8eefabd1e74d  /tmp/junit11145264397678655906/sourceCode/sub/file2
```
3. Hash the previous result as a string to get the final hash
```
3bb6a6a7b20abf850420af7fed48758f
```

The problem that exists currently is that the MD5 MessageDigest in Java produces upper case hash values, e.g.,
```
ACBD18DB4CC2F85CEDEF654FCCC4A4D8 /tmp/junit11145264397678655906/sourceCode/file1
AE3E83E2FAB3A7D8683D8EEFABD1E74D /tmp/junit11145264397678655906/sourceCode/sub/file2
```

When this string is hashed, it leads to a different value than if the hashes were lower case. Resulting in the source code annotation always having `isSatisifed=false`. 

---

Another issue with #125 is that the command is listed as:
```bash
find $PATH -type f -exec md5sum {} + | LC_ALL=C sort | md5sum | cut -d" " -f1
```
PATH is a known env variable in linux systems, it is better to use `$PWD` instead. Note that `.` will not work because it will use relative paths for the file names, which is inconsistent with the Java SDK behavior.

### PR DESCRIPTION
**Pull request by Ali-Amin**
Fix #139

* SourceCodeAnnotator hashing should now be consistent with the following command

    `find $PWD -type f -exec md5sum {} + | LC_ALL=C sort | md5sum |\
                        cut -d" " -f1`

     Where `$PWD` is the codebase directory

* Added new line and leading/trailing space handling for the
  existing checksum file
* Added example Jenkinsfile for reference that builds the SDK using
  the alvarium-pipelines lib

### CODE REVIEW

File: src/test/java/com/alvarium/annotators/SourceCodeAnnotatorTest.java [hunk 1]
```diff
@@ -92,9 +92,9 @@ public void executeShouldReturnAnnotation() throws AnnotatorException, IOExcepti
                 final File checksumFile = dir.newFile("checksum");
     
                 final HashProvider hash = new HashProviderFactory().getProvider(config.getHash().getType());
-                String hashAndPath = hash.derive(Files.readAllBytes(f1.toPath())) + "  " + f1.toPath().toString() + "\n";
-                hashAndPath = hashAndPath + hash.derive(Files.readAllBytes(f2.toPath())) + "  " + f2.toPath().toString() + "\n";
-                final String checksum = hash.derive(hashAndPath.getBytes());
+                String hashAndPath = hash.derive(Files.readAllBytes(f1.toPath())).toLowerCase() + "  " + f1.toPath().toString() + "\n";
+                hashAndPath = hashAndPath + hash.derive(Files.readAllBytes(f2.toPath())).toLowerCase() + "  " + f2.toPath().toString() + "\n";
+                final String checksum = hash.derive(hashAndPath.getBytes()).toLowerCase();
```
Comment (michaelehab) [inline src/test/java/com/alvarium/annotators/SourceCodeAnnotatorTest.java:97]: I have reviewed the recent fix made to the source-code annotator logic, it looks good, I tried it and it produces satisfied annotations, but I noticed that the corresponding test is failing. It appears that the test needs to be updated to reflect the changes in the annotator logic.
I suggest these changes:
```suggestion
                final HashProvider hash = new HashProviderFactory().getProvider(config.getHash().getType());
                List<String> hashesAndPaths = new ArrayList<>();
                hashesAndPaths.add(hash.derive(Files.readAllBytes(f1.toPath())).toLowerCase() + "  "
                                + f1.toPath().toString().replace(sourceCodeDir.toPath().toString(), "."));
                hashesAndPaths.add(hash.derive(Files.readAllBytes(f2.toPath())).toLowerCase() + "  "
                                + f2.toPath().toString().replace(sourceCodeDir.toPath().toString(), "."));
                Collections.sort(hashesAndPaths, Collator.getInstance(Locale.US));

                String hashesAndFiles = String.join("\n", hashesAndPaths) + "\n";
                final String checksum = hash.derive(hashesAndFiles.getBytes()).toLowerCase();
```
After implementing these changes, I have verified that all tests pass successfully. I have attached a screenshot for your reference.
image

**Review by michaelehab — CHANGES_REQUESTED** [review]

**Review by tsconn23 — APPROVED** [review]
LGTM pending Michael's suggestion on the test

File: src/test/java/com/alvarium/annotators/SourceCodeAnnotatorTest.java (hunk 1 above)
Comment (Ali-Amin) [reply src/test/java/com/alvarium/annotators/SourceCodeAnnotatorTest.java]: Thanks for catching that, I must've been too caught up in testing on Jenkins that I forgot to run the unit tests hahaha. 

I added your suggestion in the latest commit, just lifted some of the logic into some additional variables.

**Review by michaelehab — APPROVED** [review]
LGTM, Thanks @Ali-Amin

**Pull request merged**